File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head):
        prev = None
        curr = head
        while curr:
            next_temp = curr.next  # Temporarily store the next node
            curr.next = prev  # Reverse current node's pointer
            prev = curr  # Move pointers one step forward
            curr = next_temp
        return prev

File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null; 
        ListNode curr = head; 
        // null <- 1  2 -> 3 -> 4 -> 5
        while (curr != null) {
            ListNode temp = curr.next; 
            curr.next = prev; 
            prev = curr; 
            curr = temp; 
        }
        return prev; 
    }
}